content/blog/2023-10-18-skywalking-toolkit-trace/index.md

```diff
@@ -0,0 +1,285 @@
+	r.Header.Add(headerKey, headerValue)
+	return nil
+})
+resp, err := http.Get("http://localhost:9999/provider")
+trace.SetTag("status_code", fmt.Sprintf("%d", resp.StatusCode))
+spanID := trace.GetSpanID()
+trace.StopSpan()
+```
+
+It's important to note that when making these method calls, the current thread should have an active span.
+
+## Async APIs
+
+Async APIs work for manipulating spans across Goroutines. These scenarios might include:
+
+- Applications involving concurrency or multiple goroutines where operating on Spans across different execution contexts is necessary.
+- Updating or logging information for a Span during asynchronous operations.
+- Requiring a delayed completion of a Span.
+
+To use it, follow these steps:
+
+- Obtain the return value of CreateSpan, which is `SpanRef`.
+- Call `spanRef.PrepareAsync()` to prepare for operations in another goroutine.
+- When the current goroutine's work is done, call `trace.StopSpan()` to end the span (affecting only in the current goroutine).
+- Pass the `spanRef` to another goroutine.
+- After the work is done in any goroutine, call `spanRef.AsyncFinish()`.
+
+Here's an example:
+
+```go
+spanRef, _ := trace.CreateLocalSpan("LocalSpan")
+spanRef.PrepareAsync()
+go func(){
+	// some work
+    spanRef.AsyncFinish()
+}()
+// some work
+trace.StopSpan()
+```
+
+## Correlation Context
+
+Correlation Context is used to pass parameters within a Span, and the parent Span will pass the Correlation Context to all its child Spans. It allows the transmission of information between spans across different applications. The default number of elements in the Correlation Context is 3, and the content's length cannot exceed 128 bytes.
+
+Correlation Context is commonly applied in the following scenarios:
+
+- `Passing Information Between Spans`: It facilitates the transfer of critical information between different Spans, enabling upstream and downstream Spans to understand the correlation and context between each other.
+- `Passing Business Parameters`: In business scenarios, it involves transmitting specific parameters or information between different Spans, such as authentication tokens, business transaction IDs, and more.
+
+Users can set the Correlation Context using `trace.SetCorrelation(key, value)` and then retrieve the corresponding value in downstream spans using `value := trace.GetCorrelation(key)`.
+
+For example, in the code below, we store the value in the tag of the span, making it easier to observe the result:
+
+```go
+package main
+
+import (
+	_ "github.com/apache/skywalking-go"
+	"github.com/apache/skywalking-go/toolkit/trace"
+	"net/http"
+)
+
+func providerHandler(w http.ResponseWriter, r *http.Request) {
+	ctxValue := trace.GetCorrelation("key")
+	trace.SetTag("result", ctxValue)
+}
+
+func consumerHandler(w http.ResponseWriter, r *http.Request) {
+	trace.SetCorrelation("key", "value")
```

Comment:
What kind of key and value shall be provided for the correlation? Do we have a best practice?

Reply:
This depends on the needs of your business system. You just need to ensure that the content you want to transmit does not exceed the limits set in the configuration.

Reply:
Actually, end users would choose. Sometimes, they put order ID, product ID, etc.